In [33]:
import pandas as pd
import numpy as np  
from sklearn.linear_model import LinearRegression

In [34]:
bt = pd.read_excel("Boston Housing Dataset 1978.xlsx")
bt.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

In [35]:
import statsmodels.api as sm

nbt=len(bt)
x_multiple = [np.ones([nbt,1]),bt['CRIM'],bt['ZN'],bt['INDUS'],bt['CHAS'],bt['NOX'],bt['RM'],bt['AGE'],bt['DIS'],bt['RAD'],bt['TAX'],bt['PTRATIO'],bt['B'],bt['LSTAT']]
y = bt['MEDV'].values
X_multiple = np.column_stack(x_multiple)
ols =sm.OLS(y, X_multiple)
results = ols.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                         nan
Model:                            OLS   Adj. R-squared:                    nan
Method:                 Least Squares   F-statistic:                       nan
Date:                Mon, 01 Sep 2025   Prob (F-statistic):                nan
Time:                        17:18:49   Log-Likelihood:                    nan
No. Observations:                 506   AIC:                               nan
Df Residuals:                     492   BIC:                               nan
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const             nan        nan        nan        n

In [37]:
bt.tail(5)


CRIM   ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
501  0.06263  0.0  11.93   0.0  0.573  6.593  69.1  2.4786    1  273.0   
502  0.04527  0.0  11.93   0.0  0.573  6.120  76.7  2.2875    1  273.0   
503  0.06076  0.0  11.93   0.0  0.573  6.976  91.0  2.1675    1  273.0   
504  0.10959  0.0  11.93   0.0  0.573  6.794  89.3  2.3889    1  273.0   
505  0.04741  0.0  11.93   0.0  0.573  6.030  80.8  2.5050    1  273.0   

     PTRATIO       B  LSTAT  MEDV  
501     21.0  391.99   9.67  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9

Tenemos datos faltantes en la y (MEDV) con 452 datos y en la x tenemos 506. Tenemos que hacer una imputación o eliminar. Donde la imputación seria con media, patrones o moda. No sabemos cual de esos 3 todavia. 

In [38]:
bt_clean=bt.tail(5)
print(bt_clean)

        CRIM   ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
501  0.06263  0.0  11.93   0.0  0.573  6.593  69.1  2.4786    1  273.0   
502  0.04527  0.0  11.93   0.0  0.573  6.120  76.7  2.2875    1  273.0   
503  0.06076  0.0  11.93   0.0  0.573  6.976  91.0  2.1675    1  273.0   
504  0.10959  0.0  11.93   0.0  0.573  6.794  89.3  2.3889    1  273.0   
505  0.04741  0.0  11.93   0.0  0.573  6.030  80.8  2.5050    1  273.0   

     PTRATIO       B  LSTAT  MEDV  
501     21.0  391.99   9.67  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9  


In [39]:
bt_clean.describe()

CRIM   ZN  INDUS  CHAS    NOX        RM        AGE       DIS  RAD  \
count  5.000000  5.0   5.00   5.0  5.000  5.000000   5.000000  5.000000  5.0   
mean   0.065132  0.0  11.93   0.0  0.573  6.502600  81.380000  2.365500  1.0   
std    0.026031  0.0   0.00   0.0  0.000  0.414405   9.059636  0.139721  0.0   
min    0.045270  0.0  11.93   0.0  0.573  6.030000  69.100000  2.167500  1.0   
25%    0.047410  0.0  11.93   0.0  0.573  6.120000  76.700000  2.287500  1.0   
50%    0.060760  0.0  11.93   0.0  0.573  6.593000  80.800000  2.388900  1.0   
75%    0.062630  0.0  11.93   0.0  0.573  6.794000  89.300000  2.478600  1.0   
max    0.109590  0.0  11.93   0.0  0.573  6.976000  91.000000  2.505000  1.0   

         TAX  PTRATIO           B     LSTAT       MEDV  
count    5.0      5.0    5.000000  5.000000   5.000000  
mean   273.0     21.0  395.228000  7.750000  20.160000  
std      0.0      0.0    2.346949  1.698323   4.764767  
min    273.0     21.0  391.990000  5.640000  11.900000  
25%    273.0     21.0  393.450000  6.480000  20.600000  
50%    273.0     21.0  396.900000  7.880000  22.000000  
75%    273.0     21.0  396.900000  9.080000  22.400000  
max    273.0     21.0  396.900000  9.670000  23.900000

In [40]:
bt_clean=bt.copy()

In [41]:
bt_clean.dropna(inplace=True)

In [42]:
import numpy as np
import statsmodels.api as sm
nbt = len(bt_clean)
x_multiple = [np.ones([nbt,1]), bt_clean['CRIM'], bt_clean['ZN'], bt_clean['INDUS'], bt_clean['CHAS'], bt_clean['NOX'], bt_clean['RM'], bt_clean['AGE'], bt_clean['DIS'], bt_clean['RAD'], bt_clean['TAX'], bt_clean['PTRATIO'], bt_clean['B'], bt_clean['LSTAT']]
y = bt_clean['MEDV'].values
X_multiple = np.column_stack(x_multiple)
ols = sm.OLS(y, X_multiple)
results = ols.fit()
print(results.summary())


                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.741
Model:                            OLS   Adj. R-squared:                  0.733
Method:                 Least Squares   F-statistic:                     96.16
Date:                Mon, 01 Sep 2025   Prob (F-statistic):          2.64e-119
Time:                        17:18:56   Log-Likelihood:                -1319.4
No. Observations:                 452   AIC:                             2667.
Df Residuals:                     438   BIC:                             2724.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         20.8664      5.477      3.810      0.0

# Min max

# Escalamiento estandar